File: _147_InsertionSortList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertionSortList_list(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #Runtime: 2496 ms, faster than 9.33% of Python3 online submissions for 
        #Insertion Sort List. Memory Usage: 16.3 MB, less than 68.48% of 
        #Python3 online submissions for Insertion Sort List.
        if not head or not head.next:
            return head
        A = []
        node = head
        while node:
            A.append(node)
            node = node.next
        for j in range(1, len(A)):
            tmp = A[j]
            i = j - 1
            while i >= 0 and A[i].val > tmp.val:
                A[i + 1] = A[i]
                i -= 1
            A[i + 1] = tmp
        for index in range(len(A) - 1):
            A[index].next = A[index + 1]
        A[-1].next = None
        return A[0]

    def insertionSortList_node(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #Runtime: 188 ms, faster than 71.78% of Python3 online submissions for 
        #Insertion Sort List. Memory Usage: 16.5 MB, less than 30.28% of 
        #Python3 online submissions for Insertion Sort List.
        if not head or not head.next:
            return head
        
        dummyHead = ListNode(next=head)
        lastSorted = head
        cur = head.next
        while cur:
            if cur.val >= lastSorted.val:
                lastSorted = lastSorted.next
            else:
                prev = dummyHead
                while prev.next.val < cur.val:
                    prev = prev.next
                lastSorted.next = cur.next
                cur.next = prev.next
                prev.next = cur
            cur = lastSorted.next
        return dummyHead.next

File: _147_InsertionSortList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode curr = head;
        ListNode dummy = new ListNode();

        // We save all kinds of trouble with prev and next by a reordered
        // linkedlist
        while (curr != null) {
            ListNode prev = dummy;
            while (prev.next != null && prev.next.val <= curr.val) {
                prev = prev.next;
            }
            ListNode nxt = curr.next;
            curr.next = prev.next;
            prev.next = curr;
            curr = nxt;
        }

        return dummy.next;
    }
}